File: challenges/Algorithmic/MBR/mbr.py
"""mbr.py - Educational Master Boot Record (MBR) partition table parser.

This module provides a small, well‑documented parser for the 512‑byte
Master Boot Record sector found at the beginning of traditional MBR
partitioned disks. It focuses on clarity and approachability for new
developers while still demonstrating robust, testable design.

Features
========
* Dataclass models for partition entries and overall MBR parse result
* Safe parsing with validation & helpful error messages
* CHS (Cylinder / Head / Sector) decoding (legacy – informational only)
* Human friendly size formatting and optional JSON output
* Simple CLI built with argparse
* Utility to create a reproducible dummy MBR for demonstrations / tests

Example (human readable):
    python mbr.py --file disk.img

Example (JSON):
    python mbr.py --file disk.img --json

Create and parse a dummy image:
    python mbr.py --create-dummy demo_mbr.bin --json

Programmatic usage:
    from mbr import read_mbr, parse_mbr
    raw = read_mbr('disk.img')
    result = parse_mbr(raw)
    for p in result.partitions:
        print(p.index, p.type_description, p.size_human)

This code intentionally stays within the Python standard library.
"""

from __future__ import annotations

import argparse
import json
import os
import struct
import sys
from dataclasses import dataclass
from typing import Dict, Iterable, List, Optional, Tuple

SECTOR_BYTES_DEFAULT = 512
MBR_SIZE = 512
PARTITION_TABLE_OFFSET = 446
PARTITION_ENTRY_SIZE = 16
PARTITION_ENTRY_COUNT = 4
MBR_SIGNATURE_OFFSET = 510
MBR_SIGNATURE = 0xAA55

# Common partition type codes (not exhaustive – extend as needed)
PARTITION_TYPE_MAP: Dict[int, str] = {
    0x00: "Unused",
    0x01: "FAT12",
    0x04: "FAT16 <32M",
    0x05: "Extended",
    0x06: "FAT16",
    0x07: "NTFS/ExFAT/HPFS",
    0x0B: "FAT32",
    0x0C: "FAT32 LBA",
    0x0E: "FAT16 LBA",
    0x0F: "Extended LBA",
    0x82: "Linux Swap",
    0x83: "Linux Filesystem",
    0x84: "Hibernate",
    0x85: "Linux Extended",
    0x8E: "Linux LVM",
    0xA5: "FreeBSD",
    0xA8: "macOS UFS",
    0xAB: "macOS Boot",
    0xAF: "macOS HFS/HFS+",
    0xEE: "GPT Protective",
    0xEF: "EFI System",
}


def _decode_chs(triple: bytes) -> Tuple[int, int, int]:
    """Decode 3 CHS bytes into (cylinder, head, sector).

    CHS layout (legacy):
        Byte 0: Head (0–255)
        Byte 1: Sector (bits 0–5, 1–63) | high 2 bits of Cylinder (bits 6–7)
        Byte 2: Low 8 bits of Cylinder
    """
    if len(triple) != 3:
        return (0, 0, 0)
    head = triple[0]
    sector = triple[1] & 0x3F  # lower 6 bits
    cylinder = ((triple[1] & 0xC0) << 2) | triple[2]
    return (cylinder, head, sector)


def _human_size(num_bytes: int) -> str:
    """Return a human friendly size string using binary prefixes."""
    if num_bytes <= 0:
        return "0 B"
    units = ["B", "KiB", "MiB", "GiB", "TiB"]
    value = float(num_bytes)
    for unit in units:
        if value < 1024 or unit == units[-1]:
            return f"{value:.2f} {unit}"
        value /= 1024
    return f"{value:.2f} {units[-1]}"  # fallback (unreachable)


@dataclass(slots=True)
class PartitionEntry:
    """Represents a single 16‑byte MBR partition table entry."""

    index: int
    bootable: bool
    status_raw: int
    type_code: int
    start_chs: Tuple[int, int, int]
    end_chs: Tuple[int, int, int]
    start_lba: int
    sectors: int
    sector_bytes: int = SECTOR_BYTES_DEFAULT

    @property
    def size_bytes(self) -> int:
        return self.sectors * self.sector_bytes

    @property
    def size_human(self) -> str:
        return _human_size(self.size_bytes)

    @property
    def type_description(self) -> str:
        return PARTITION_TYPE_MAP.get(self.type_code, "Unknown")

    def to_dict(self) -> Dict[str, object]:
        return {
            "index": self.index,
            "bootable": self.bootable,
            "status_raw": f"0x{self.status_raw:02x}",
            "type_code": f"0x{self.type_code:02x}",
            "type_description": self.type_description,
            "start_chs": {
                "cylinder": self.start_chs[0],
                "head": self.start_chs[1],
                "sector": self.start_chs[2],
            },
            "end_chs": {
                "cylinder": self.end_chs[0],
                "head": self.end_chs[1],
                "sector": self.end_chs[2],
            },
            "start_lba": self.start_lba,
            "sectors": self.sectors,
            "size_bytes": self.size_bytes,
            "size_human": self.size_human,
        }


@dataclass(slots=True)
class MBRParseResult:
    """Container for overall MBR parse outcome."""

    signature: int
    signature_valid: bool
    partitions: List[PartitionEntry]
    sector_bytes: int = SECTOR_BYTES_DEFAULT

    def to_dict(self) -> Dict[str, object]:
        return {
            "signature": f"0x{self.signature:04x}",
            "signature_valid": self.signature_valid,
            "sector_bytes": self.sector_bytes,
            "partitions": [p.to_dict() for p in self.partitions],
        }


def parse_partition_entry(
    entry_bytes: bytes, index: int, sector_bytes: int = SECTOR_BYTES_DEFAULT
) -> Optional[PartitionEntry]:
    """Parse a single 16‑byte partition entry.

    Returns None for an all‑zero (unused) entry.
    """
    if len(entry_bytes) != PARTITION_ENTRY_SIZE:
        raise ValueError("Partition entry must be 16 bytes")
    if all(b == 0 for b in entry_bytes):
        return None

    status = entry_bytes[0]
    start_chs = _decode_chs(entry_bytes[1:4])
    type_code = entry_bytes[4]
    end_chs = _decode_chs(entry_bytes[5:8])
    start_lba, sectors = struct.unpack_from("<II", entry_bytes, 8)

    return PartitionEntry(
        index=index,
        bootable=(status == 0x80),
        status_raw=status,
        type_code=type_code,
        start_chs=start_chs,
        end_chs=end_chs,
        start_lba=start_lba,
        sectors=sectors,
        sector_bytes=sector_bytes,
    )


def parse_mbr(
    mbr_data: bytes, *, sector_bytes: int = SECTOR_BYTES_DEFAULT
) -> MBRParseResult:
    """Parse raw 512‑byte MBR sector into structured result."""
    if len(mbr_data) != MBR_SIZE:
        raise ValueError(f"MBR must be exactly {MBR_SIZE} bytes; got {len(mbr_data)}")

    signature = struct.unpack_from("<H", mbr_data, MBR_SIGNATURE_OFFSET)[0]
    signature_valid = signature == MBR_SIGNATURE

    table = mbr_data[
        PARTITION_TABLE_OFFSET : PARTITION_TABLE_OFFSET
        + PARTITION_ENTRY_COUNT * PARTITION_ENTRY_SIZE
    ]
    partitions: List[PartitionEntry] = []
    for i in range(PARTITION_ENTRY_COUNT):
        start = i * PARTITION_ENTRY_SIZE
        entry_bytes = table[start : start + PARTITION_ENTRY_SIZE]
        part = parse_partition_entry(entry_bytes, i + 1, sector_bytes)
        if part is not None:
            partitions.append(part)

    return MBRParseResult(
        signature=signature,
        signature_valid=signature_valid,
        partitions=partitions,
        sector_bytes=sector_bytes,
    )


def read_mbr(path: str) -> bytes:
    """Read 512 bytes from a file path.

    Raises FileNotFoundError / IOError on failure.
    """
    with open(path, "rb") as f:
        data = f.read(MBR_SIZE)
    if len(data) != MBR_SIZE:
        raise IOError(
            f"File '{path}' shorter than {MBR_SIZE} bytes; cannot be a full MBR sector"
        )
    return data


def create_dummy_mbr_file(filepath: str, *, overwrite: bool = False) -> None:
    """Create a deterministic dummy MBR with two sample partitions.

    Partition layout:
        #1 bootable FAT32 LBA starting at LBA 2048, size 1,000,000 sectors (~476.84 MiB)
        #2 non‑bootable Linux filesystem starting at 1,002,048, size 4,000,000 sectors (~1.91 GiB)
    """
    if os.path.exists(filepath) and not overwrite:
        raise FileExistsError(
            f"File '{filepath}' already exists. Pass --force to overwrite."
        )
    mbr = bytearray(MBR_SIZE)

    # Helper to build a partition entry; CHS fields left zero for simplicity.
    def build(status: int, type_code: int, start_lba: int, sectors: int) -> bytes:
        return struct.pack(
            "<B3sB3sII",
            status,
            b"\x00\x00\x00",
            type_code,
            b"\x00\x00\x00",
            start_lba,
            sectors,
        )

    part1 = build(0x80, 0x0C, 2048, 1_000_000)
    part2 = build(0x00, 0x83, 1_002_048, 4_000_000)

    mbr[PARTITION_TABLE_OFFSET : PARTITION_TABLE_OFFSET + PARTITION_ENTRY_SIZE] = part1
    mbr[
        PARTITION_TABLE_OFFSET
        + PARTITION_ENTRY_SIZE : PARTITION_TABLE_OFFSET
        + 2 * PARTITION_ENTRY_SIZE
    ] = part2
    mbr[MBR_SIGNATURE_OFFSET : MBR_SIGNATURE_OFFSET + 2] = struct.pack(
        "<H", MBR_SIGNATURE
    )

    with open(filepath, "wb") as f:
        f.write(mbr)


def _print_human(result: MBRParseResult, *, show_empty: bool = False) -> None:
    print(
        f"MBR Signature: 0x{result.signature:04x} {'(valid)' if result.signature_valid else '(INVALID)'}"
    )
    if not result.partitions and not show_empty:
        print("No partitions present.")
        return
    print("\nPartition Table:")
    if not result.partitions and show_empty:
        for i in range(1, PARTITION_ENTRY_COUNT + 1):
            print(f"  [Entry {i}] <empty>")
        return
    for p in result.partitions:
        print(
            f"  [Entry {p.index}] {'BOOT ' if p.bootable else '     '}Type {p.type_description} ({p.type_code:#04x})"
        )
        print(
            f"      Start LBA: {p.start_lba:,}  Sectors: {p.sectors:,}  Size: {p.size_human}"
        )
        cyl_s, head_s, sec_s = p.start_chs
        cyl_e, head_e, sec_e = p.end_chs
        if any(
            [cyl_s, head_s, sec_s, cyl_e, head_e, sec_e]
        ):  # only show if not all zero
            print(
                f"      CHS: start (C={cyl_s} H={head_s} S={sec_s}) end (C={cyl_e} H={head_e} S={sec_e})"
            )


def build_arg_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Parse an MBR (Master Boot Record) sector."
    )
    parser.add_argument(
        "--file",
        "-f",
        help="Path to disk image containing an MBR (default: dummy_mbr.bin)",
    )
    parser.add_argument(
        "--json", action="store_true", help="Output JSON instead of human readable text"
    )
    parser.add_argument(
        "--show-empty",
        action="store_true",
        help="Show empty entries explicitly in human readable output",
    )
    parser.add_argument(
        "--create-dummy",
        metavar="PATH",
        help="Create a dummy MBR at PATH and exit (unless --file also used to parse)",
    )
    parser.add_argument(
        "--force",
        action="store_true",
        help="Overwrite when using --create-dummy if file exists",
    )
    parser.add_argument(
        "--sector-size",
        type=int,
        default=SECTOR_BYTES_DEFAULT,
        help="Logical sector size in bytes (default 512)",
    )
    return parser


def main(argv: Optional[Iterable[str]] = None) -> int:
    parser = build_arg_parser()
    args = parser.parse_args(list(argv) if argv is not None else None)

    # Optional dummy creation
    if args.create_dummy:
        try:
            create_dummy_mbr_file(args.create_dummy, overwrite=args.force)
            print(f"Created dummy MBR at '{args.create_dummy}'", file=sys.stderr)
        except FileExistsError as e:
            print(e, file=sys.stderr)
            return 1
        # If only creating dummy and no file specified, parse that
        if not args.file:
            args.file = args.create_dummy

    path = args.file or "dummy_mbr.bin"
    if not os.path.exists(path):
        # Auto-generate dummy if using default path and absent.
        if path == "dummy_mbr.bin":
            try:
                create_dummy_mbr_file(path, overwrite=True)
                print("Generated default dummy_mbr.bin", file=sys.stderr)
            except Exception as e:  # pragma: no cover - very unlikely
                print(f"Failed to create dummy MBR: {e}", file=sys.stderr)
                return 1
        else:
            print(f"File not found: {path}", file=sys.stderr)
            return 1

    try:
        raw = read_mbr(path)
        result = parse_mbr(raw, sector_bytes=args.sector_size)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        return 1

    if args.json:
        print(json.dumps(result.to_dict(), indent=2))
    else:
        _print_human(result, show_empty=args.show_empty)
    return 0 if result.signature_valid else 2


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(main())

File: challenges/Algorithmic/MBR/MBR.hs
{-# LANGUAGE OverloadedStrings #-}

-- |
-- Module      : Main
-- Description : Educational Master Boot Record parser implemented in Haskell.
--
-- This standalone tool mirrors the Python reference implementation by reading
-- a 512-byte sector, decoding legacy partition entries, and emitting either a
-- human readable summary or JSON. It also shares the deterministic dummy image
-- generator so both languages stay interoperable with the bundled visualiser.
module Main (main) where

import Control.Monad (foldM, unless, when)
import Data.Bits ((.&.), shiftL, shiftR)
import qualified Data.ByteString as BS
import Data.Char (toLower)
import Data.List (intercalate)
import Data.Maybe (catMaybes, fromMaybe)
import Data.Word (Word16, Word32, Word8)
import Numeric (showHex)
import System.Console.GetOpt
import System.Directory (doesFileExist)
import System.Environment (getArgs)
import System.Exit (ExitCode (..), exitFailure, exitWith)
import System.IO (hPutStrLn, stderr)
import Text.Printf (printf)

--------------------------------------------------------------------------------
-- Constants and data models ---------------------------------------------------
--------------------------------------------------------------------------------

sectorBytesDefault :: Int
sectorBytesDefault = 512

mbrSize :: Int
mbrSize = 512

partitionTableOffset :: Int
partitionTableOffset = 446

partitionEntrySize :: Int
partitionEntrySize = 16

partitionEntryCount :: Int
partitionEntryCount = 4

mbrSignatureOffset :: Int
mbrSignatureOffset = 510

mbrSignature :: Word16
mbrSignature = 0xAA55

data LogLevel
  = LogError
  | LogWarn
  | LogInfo
  | LogDebug
  deriving (Eq, Ord, Enum, Bounded, Show)

data Options = Options
  { optFile :: Maybe FilePath
  , optJson :: Bool
  , optShowEmpty :: Bool
  , optCreateDummy :: Maybe FilePath
  , optForce :: Bool
  , optSectorSize :: Int
  , optLogLevel :: LogLevel
  }

defaultOptions :: Options
defaultOptions =
  Options
    { optFile = Nothing
    , optJson = False
    , optShowEmpty = False
    , optCreateDummy = Nothing
    , optForce = False
    , optSectorSize = sectorBytesDefault
    , optLogLevel = LogWarn
    }

data PartitionEntry = PartitionEntry
  { partIndex :: Int
  , partBootable :: Bool
  , partStatusRaw :: Word8
  , partTypeCode :: Word8
  , partStartCHS :: (Int, Int, Int)
  , partEndCHS :: (Int, Int, Int)
  , partStartLBA :: Word32
  , partSectors :: Word32
  , partSectorBytes :: Int
  }
  deriving (Show)

data MBRParseResult = MBRParseResult
  { mbrSignatureValue :: Word16
  , mbrSignatureValid :: Bool
  , mbrPartitions :: [PartitionEntry]
  , mbrSectorBytes :: Int
  }
  deriving (Show)

--------------------------------------------------------------------------------
-- Logging ---------------------------------------------------------------------
--------------------------------------------------------------------------------

parseLogLevel :: String -> Either String LogLevel
parseLogLevel raw =
  case map toLower raw of
    "error" -> Right LogError
    "warn" -> Right LogWarn
    "warning" -> Right LogWarn
    "info" -> Right LogInfo
    "debug" -> Right LogDebug
    _ -> Left $ "Unknown log level '" ++ raw ++ "' (expected error|warn|info|debug)"

logMessage :: LogLevel -> Options -> String -> IO ()
logMessage level opts msg =
  when (level <= optLogLevel opts) $ do
    let prefix = case level of
          LogError -> "[ERROR] "
          LogWarn -> "[WARN ] "
          LogInfo -> "[INFO ] "
          LogDebug -> "[DEBUG] "
    hPutStrLn stderr (prefix ++ msg)

--------------------------------------------------------------------------------
-- CLI -------------------------------------------------------------------------
--------------------------------------------------------------------------------

optionsSpec :: [OptDescr (Options -> Either String Options)]
optionsSpec =
  [ Option ['f'] ["file"] (ReqArg setFile "PATH") "Path to disk image containing an MBR (default: dummy_mbr.bin)"
  , Option [] ["json"] (NoArg setJson) "Output JSON instead of human readable text"
  , Option [] ["show-empty"] (NoArg setShowEmpty) "Show empty entries explicitly in human readable mode"
  , Option [] ["create-dummy"] (ReqArg setCreate "PATH") "Create a deterministic dummy MBR at PATH"
  , Option [] ["force"] (NoArg setForce) "Overwrite when using --create-dummy"
  , Option [] ["sector-size"] (ReqArg setSector "BYTES") "Logical sector size in bytes (default 512)"
  , Option [] ["log"] (ReqArg setLog "LEVEL") "Set logging verbosity (error|warn|info|debug)"
  ]
  where
    setFile path opts = Right $ opts {optFile = Just path}
    setJson opts = Right $ opts {optJson = True}
    setShowEmpty opts = Right $ opts {optShowEmpty = True}
    setCreate path opts = Right $ opts {optCreateDummy = Just path}
    setForce opts = Right $ opts {optForce = True}
    setSector arg opts =
      case reads arg of
        [(n, "")] | n > 0 -> Right $ opts {optSectorSize = n}
        _ -> Left $ "Invalid sector size '" ++ arg ++ "'"
    setLog arg opts =
      case parseLogLevel arg of
        Right lvl -> Right $ opts {optLogLevel = lvl}
        Left err -> Left err

parseOptions :: [String] -> IO Options
parseOptions argv = do
  let (actions, nonOpts, errs) = getOpt Permute optionsSpec argv
  unless (null errs) $ do
    mapM_ (hPutStrLn stderr) errs
    hPutStrLn stderr (usageInfo "Usage: mbr [OPTIONS]" optionsSpec)
    exitFailure
  unless (null nonOpts) $ do
    hPutStrLn stderr $ "Unexpected positional arguments: " ++ unwords nonOpts
    hPutStrLn stderr (usageInfo "Usage: mbr [OPTIONS]" optionsSpec)
    exitFailure
  foldM apply defaultOptions actions
  where
    apply opts action =
      case action opts of
        Right updated -> pure updated
        Left err -> do
          hPutStrLn stderr err
          hPutStrLn stderr (usageInfo "Usage: mbr [OPTIONS]" optionsSpec)
          exitFailure

--------------------------------------------------------------------------------
-- Partition helpers -----------------------------------------------------------
--------------------------------------------------------------------------------

typeMap :: [(Word8, String)]
typeMap =
  [ (0x00, "Unused")
  , (0x01, "FAT12")
  , (0x04, "FAT16 <32M")
  , (0x05, "Extended")
  , (0x06, "FAT16")
  , (0x07, "NTFS/ExFAT/HPFS")
  , (0x0B, "FAT32")
  , (0x0C, "FAT32 LBA")
  , (0x0E, "FAT16 LBA")
  , (0x0F, "Extended LBA")
  , (0x82, "Linux Swap")
  , (0x83, "Linux Filesystem")
  , (0x84, "Hibernate")
  , (0x85, "Linux Extended")
  , (0x8E, "Linux LVM")
  , (0xA5, "FreeBSD")
  , (0xA8, "macOS UFS")
  , (0xAB, "macOS Boot")
  , (0xAF, "macOS HFS/HFS+")
  , (0xEE, "GPT Protective")
  , (0xEF, "EFI System")
  ]

typeDescription :: Word8 -> String
typeDescription code = fromMaybe "Unknown" (lookup code typeMap)

decodeCHS :: BS.ByteString -> (Int, Int, Int)
decodeCHS triple
  | BS.length triple /= 3 = (0, 0, 0)
  | otherwise =
      let h = fromIntegral (BS.index triple 0)
          s = fromIntegral (BS.index triple 1 .&. 0x3F)
          cylHigh = fromIntegral ((BS.index triple 1 .&. 0xC0) `shiftL` 2)
          cylLow = fromIntegral (BS.index triple 2)
          c = cylHigh + cylLow
       in (c, h, s)

leWord32 :: BS.ByteString -> Word32
leWord32 bs =
  foldl
    (\acc (i, b) -> acc + (fromIntegral b `shiftL` (8 * i)))
    (0 :: Word32)
    (zip [0 .. 3] (map (BS.index bs) [0 .. 3]))

leWord16 :: BS.ByteString -> Word16
leWord16 bs =
  foldl
    (\acc (i, b) -> acc + (fromIntegral b `shiftL` (8 * i)))
    (0 :: Word16)
    (zip [0 .. 1] (map (BS.index bs) [0 .. 1]))

parsePartitionEntry :: BS.ByteString -> Int -> Int -> Maybe PartitionEntry
parsePartitionEntry entry idx sectorBytes
  | BS.length entry /= partitionEntrySize = Nothing
  | all (== 0) (BS.unpack entry) = Nothing
  | otherwise =
      let status = BS.index entry 0
          startChs = decodeCHS (BS.take 3 (BS.drop 1 entry))
          typeCode = BS.index entry 4
          endChs = decodeCHS (BS.take 3 (BS.drop 5 entry))
          startLba = leWord32 (BS.take 4 (BS.drop 8 entry))
          sectors = leWord32 (BS.take 4 (BS.drop 12 entry))
       in Just
            PartitionEntry
              { partIndex = idx
              , partBootable = status == 0x80
              , partStatusRaw = status
              , partTypeCode = typeCode
              , partStartCHS = startChs
              , partEndCHS = endChs
              , partStartLBA = startLba
              , partSectors = sectors
              , partSectorBytes = sectorBytes
              }

parseMBR :: BS.ByteString -> Int -> Either String MBRParseResult
parseMBR raw sectorBytes
  | BS.length raw /= mbrSize = Left $ "MBR must be exactly " ++ show mbrSize ++ " bytes"
  | otherwise =
      let signature = leWord16 (BS.drop mbrSignatureOffset raw)
          signatureValid = signature == mbrSignature
          table = BS.take (partitionEntryCount * partitionEntrySize) (BS.drop partitionTableOffset raw)
          entries =
            [ parsePartitionEntry (BS.take partitionEntrySize (BS.drop (i * partitionEntrySize) table)) (i + 1) sectorBytes
            | i <- [0 .. partitionEntryCount - 1]
            ]
       in Right
            MBRParseResult
              { mbrSignatureValue = signature
              , mbrSignatureValid = signatureValid
              , mbrPartitions = catMaybes entries
              , mbrSectorBytes = sectorBytes
              }

--------------------------------------------------------------------------------
-- Rendering -------------------------------------------------------------------
--------------------------------------------------------------------------------

sizeBytes :: PartitionEntry -> Integer
sizeBytes part = fromIntegral (partSectors part) * fromIntegral (partSectorBytes part)

humanSize :: Integer -> String
humanSize n
  | n <= 0 = "0 B"
  | otherwise = go (fromIntegral n :: Double) units
  where
    units = ["B", "KiB", "MiB", "GiB", "TiB", "PiB"]
    go value (u : rest)
      | value < 1024 || null rest = printf "%.2f %s" value u
      | otherwise = go (value / 1024) rest
    go value [] = printf "%.2f TiB" value

formatHex :: (Integral a) => a -> String
formatHex v = "0x" ++ pad (map toLower (showHex v ""))
  where
    targetWidth
      | v < 0x100 = 2
      | v < 0x10000 = 4
      | otherwise = length (showHex v "")
    pad hexDigits = replicate (targetWidth - length hexDigits) '0' ++ hexDigits

showWithCommas :: (Integral a, Show a) => a -> String
showWithCommas n =
  let s = show n
      groups = reverse (chunk (reverse s))
   in intercalate "," (map reverse groups)
  where
    chunk [] = []
    chunk xs = take 3 xs : chunk (drop 3 xs)

printHuman :: Options -> MBRParseResult -> IO ()
printHuman opts result = do
  putStrLn $ "MBR Signature: " ++ formatHex (mbrSignatureValue result) ++ if mbrSignatureValid result then " (valid)" else " (INVALID)"
  let parts = mbrPartitions result
  if null parts && not (optShowEmpty opts)
    then putStrLn "No partitions present."
    else do
      putStrLn ""
      putStrLn "Partition Table:"
      if null parts && optShowEmpty opts
        then mapM_ (\i -> putStrLn ("  [Entry " ++ show i ++ "] <empty>")) [1 .. partitionEntryCount]
        else mapM_ printPartition parts
  where
    printPartition part = do
      let boot = if partBootable part then "BOOT " else "     "
          desc = typeDescription (partTypeCode part)
          typeCodeStr = formatHex (partTypeCode part)
      putStrLn $ "  [Entry " ++ show (partIndex part) ++ "] " ++ boot ++ "Type " ++ desc ++ " (" ++ typeCodeStr ++ ")"
      putStrLn $
        "      Start LBA: "
          ++ showWithCommas (partStartLBA part)
          ++ "  Sectors: "
          ++ showWithCommas (partSectors part)
          ++ "  Size: "
          ++ humanSize (sizeBytes part)
      let (cylS, headS, secS) = partStartCHS part
          (cylE, headE, secE) = partEndCHS part
      when (any (/= 0) [cylS, headS, secS, cylE, headE, secE]) $ do
        putStrLn $
          "      CHS: start (C="
            ++ show cylS
            ++ " H="
            ++ show headS
            ++ " S="
            ++ show secS
            ++ ") end (C="
            ++ show cylE
            ++ " H="
            ++ show headE
            ++ " S="
            ++ show secE
            ++ ")"

jsonString :: String -> String
jsonString s = "\"" ++ concatMap escape s ++ "\""
  where
    escape '"' = "\\\""
    escape '\\' = "\\\\"
    escape '\n' = "\\n"
    escape '\r' = "\\r"
    escape '\t' = "\\t"
    escape c = [c]

partitionToJson :: PartitionEntry -> [String]
partitionToJson part =
  [ "    {"
  , "      \"index\": " ++ show (partIndex part) ++ ","
  , "      \"bootable\": " ++ map toLower (show (partBootable part)) ++ ","
  , "      \"status_raw\": " ++ jsonString (formatHex (partStatusRaw part)) ++ ","
  , "      \"type_code\": " ++ jsonString (formatHex (partTypeCode part)) ++ ","
  , "      \"type_description\": " ++ jsonString (typeDescription (partTypeCode part)) ++ ","
  , "      \"start_chs\": {"
  , "        \"cylinder\": " ++ show (cyl (partStartCHS part)) ++ ","
  , "        \"head\": " ++ show (hed (partStartCHS part)) ++ ","
  , "        \"sector\": " ++ show (sec (partStartCHS part))
  , "      },"
  , "      \"end_chs\": {"
  , "        \"cylinder\": " ++ show (cyl (partEndCHS part)) ++ ","
  , "        \"head\": " ++ show (hed (partEndCHS part)) ++ ","
  , "        \"sector\": " ++ show (sec (partEndCHS part))
  , "      },"
  , "      \"start_lba\": " ++ show (partStartLBA part) ++ ","
  , "      \"sectors\": " ++ show (partSectors part) ++ ","
  , "      \"size_bytes\": " ++ show (sizeBytes part) ++ ","
  , "      \"size_human\": " ++ jsonString (humanSize (sizeBytes part))
  , "    }"
  ]
  where
    cyl (c, _, _) = c
    hed (_, h, _) = h
    sec (_, _, s) = s

resultToJson :: MBRParseResult -> String
resultToJson result =
  let header =
        [ "{"
        , "  \"signature\": " ++ show (mbrSignatureValue result) ++ ","
        , "  \"signature_valid\": " ++ map toLower (show (mbrSignatureValid result)) ++ ","
        , "  \"sector_bytes\": " ++ show (mbrSectorBytes result) ++ ","
        ]
      blocks = map partitionToJson (mbrPartitions result)
      partitionLines =
        case blocks of
          [] -> ["  \"partitions\": []"]
          _ ->
            ["  \"partitions\": ["]
              ++ concat
                ( zipWith
                    (\idx block -> (if idx > 0 then ["  ,"] else []) ++ block)
                    [0 :: Int ..]
                    blocks
                )
              ++ ["  ]"]
      footer = ["}"]
   in intercalate "\n" (header ++ partitionLines ++ footer)

--------------------------------------------------------------------------------
-- Dummy image -----------------------------------------------------------------
--------------------------------------------------------------------------------

createDummy :: Options -> FilePath -> IO ()
createDummy opts path = do
  exists <- doesFileExist path
  if exists && not (optForce opts)
    then do
      logMessage LogError opts $ "File '" ++ path ++ "' already exists (use --force to overwrite)"
      exitFailure
    else do
      logMessage LogInfo opts $ "Writing dummy MBR to '" ++ path ++ "'"
      let entry status typeCode startLba sectors =
            BS.concat
              [ BS.pack [status, 0, 0, 0, typeCode, 0, 0, 0]
              , le32 startLba
              , le32 sectors
              ]
          part1 = entry 0x80 0x0C 2048 1000000
          part2 = entry 0x00 0x83 1002048 4000000
          emptyEntries = BS.replicate (partitionEntrySize * 2) 0
          prefix = BS.replicate partitionTableOffset 0
          signatureBytes = BS.pack [lowByte mbrSignature, highByte mbrSignature]
          mbr = BS.concat [prefix, part1, part2, emptyEntries, signatureBytes]
      BS.writeFile path mbr
      logMessage LogInfo opts "Dummy image created"
  where
    lowByte :: Word16 -> Word8
    lowByte v = fromIntegral (v .&. 0x00FF)
    highByte :: Word16 -> Word8
    highByte v = fromIntegral ((v `shiftR` 8) .&. 0x00FF)
    le32 :: Word32 -> BS.ByteString
    le32 v =
      BS.pack
        [ fromIntegral (v .&. 0xFF)
        , fromIntegral ((v `shiftR` 8) .&. 0xFF)
        , fromIntegral ((v `shiftR` 16) .&. 0xFF)
        , fromIntegral ((v `shiftR` 24) .&. 0xFF)
        ]

--------------------------------------------------------------------------------
-- File IO ---------------------------------------------------------------------
--------------------------------------------------------------------------------

readMBR :: FilePath -> IO BS.ByteString
readMBR path = do
  raw <- BS.readFile path
  if BS.length raw < mbrSize
    then ioError (userError $ "File '" ++ path ++ "' shorter than " ++ show mbrSize ++ " bytes")
    else pure (BS.take mbrSize raw)

--------------------------------------------------------------------------------
-- Main ------------------------------------------------------------------------
--------------------------------------------------------------------------------

main :: IO ()
main = do
  argv <- getArgs
  optsBase <- parseOptions argv
  opts <-
    case optCreateDummy optsBase of
      Just target -> do
        createDummy optsBase target
        let newFile = Just (fromMaybe target (optFile optsBase))
        pure optsBase {optFile = newFile, optCreateDummy = Nothing}
      Nothing -> pure optsBase
  runWithDefault opts

runWithDefault :: Options -> IO ()
runWithDefault opts = do
  let path = fromMaybe "dummy_mbr.bin" (optFile opts)
  exists <- doesFileExist path
  unless exists $ do
    if path == "dummy_mbr.bin"
      then do
        logMessage LogInfo opts "Generating default dummy_mbr.bin"
        let forced = opts {optForce = True}
        createDummy forced path
      else do
        logMessage LogError opts $ "File not found: " ++ path
        exitFailure
  runWithFile opts path

runWithFile :: Options -> FilePath -> IO ()
runWithFile opts path = do
  logMessage LogInfo opts $ "Reading MBR from '" ++ path ++ "'"
  raw <- readMBR path
  case parseMBR raw (optSectorSize opts) of
    Left err -> do
      logMessage LogError opts err
      exitFailure
    Right result -> do
      logMessage LogDebug opts $ "Decoded " ++ show (length (mbrPartitions result)) ++ " partition entries"
      if optJson opts
        then putStrLn (resultToJson result)
        else printHuman opts result
      let exitCode = if mbrSignatureValid result then ExitSuccess else ExitFailure 2
      exitWith exitCode
